bee-protocol/src/workers/index_updater.rs

```diff
@@ -0,0 +1,216 @@
+        if index > tangle.get_latest_milestone_index() - tangle.config().below_max_depth() {
+            update_future_cone(tangle, roots).await;
+
+            // Update tip pool after all values got updated.
+            tangle.update_tip_scores().await;
+        }
+    }
+}
+
+async fn update_past_cone<B: StorageBackend>(
+    tangle: &Tangle<B>,
+    mut parents: Vec<MessageId>,
+    index: MilestoneIndex,
+) -> HashSet<MessageId> {
+    let mut updated = HashSet::new();
+
+    while let Some(parent_id) = parents.pop() {
+        // Our skip conditions:
+        // 1) check if we already updated it during this run
+        // 2) check if it's a SEP
+        // 3) check if we already updated it during a previous run
+        // Note that the order of calls is important (from cheap to more expensive) for performance reasons.
+        if updated.contains(&parent_id)
+            || tangle.is_solid_entry_point(&parent_id).await
+            || tangle
+                .get_metadata(&parent_id)
+                .await
+                // TODO: I don't think unwrapping here is safe. Investigate!
+                .unwrap()
+                .milestone_index()
+                .is_some()
+        {
+            continue;
+        }
+
+        tangle
+            .update_metadata(&parent_id, |metadata| {
+                metadata.set_milestone_index(index);
+                // TODO: That was fine in a synchronous scenario, where this algo had the newest information,
+                // but probably isn't the case in the now asynchronous scenario. Investigate!
+                metadata.set_omrsi(IndexId::new(index, parent_id));
+                metadata.set_ymrsi(IndexId::new(index, parent_id));
+            })
+            .await;
+
+        if let Some(parent) = tangle.get(&parent_id).await {
+            parents.extend_from_slice(parent.parents())
+        }
+
+        // Preferably we would only collect the 'root messages/transactions'. They are defined as being confirmed by
+        // a milestone, but at least one of their children is not confirmed yet. One can think of them as an attachment
+        // point for new messages to the main tangle. It is ensured however, that this set *contains* the root messages
+        // as well, and during the future walk we will skip already confirmed children, which shouldn't be a performance
+        // issue.
+        updated.insert(parent_id);
+    }
+
+    debug!("Set milestone index {} to {} messages.", index, updated.len());
+
+    updated
+}
+
+// NOTE: so once a milestone comes in we have to walk the future cones of the root transactions and update their
+// OMRSI and YMRSI; during that time we need to block the propagator, otherwise it will propagate outdated data.
+async fn update_future_cone<B: StorageBackend>(tangle: &Tangle<B>, roots: HashSet<MessageId>) {
+    let mut to_process = roots.into_iter().collect::<Vec<_>>();
```

Comment:
Can't we pass the roots as a `Vec` directly?

Reply:
long time ago that I wrote this. will check

Reply:
I guess you are right. We should do the conversion into the `Vec` between `update_past_cone` and `update_future_cone`, and not here.